### PR TITLE
Refactor of initial_stress in TensorMechanics Materials

### PR DESCRIPTION
**Pull request by WilkAndy**
The overall point here is to allow time-varying elasticity tensors when initial_stress is not zero.  Presently we demand that users supply an initial strain (eigenstrain) in such cases, but that is very inconvenient in practice.
ComputeStressBase::addQpInitialStress(stress) adds the initial stress, if it is provided to stress.  As noted in the documentation, this method should be called whenever stress is computed without reference to stress_old.

At the same time, I've removed some constant_in_time stuff that i believe is now redundant, but this needs checking.  (I would also like to remove the elasticity_tensor_is_isotropic stuff that appears in the same places, but i vaugely remember that that's necessary for some obscure Bison reason!)

These changes are apparently minor, but i'd like @sapitts and @bwspenc to review just in case they can think of: a better way of doing this; or whether this breaks some design i've forgotten about.

Also some minor re-workings of some gold files and
cosserat_mc_wp.i better doco

Refs #9749

### CODE REVIEW

**Comment by WilkAndy** [comment]
This probably broke Bison, boohoo

**Comment by WilkAndy** [comment]
Could @sapitts or @bwspenc (or someone else) comment on why this broke Bison?

a

**Comment by bwspenc** [comment]
@WilkAndy There are 4 BISON tests that are running differently and/or failing to converge due to these changes.  I just looked at one, and it is pretty simple: it just uses ComputeFiniteStrainElasticStress and a constant elasticity tensor. The only thing interesting about it is that it uses a temperature-dependent thermal expansion eigenstrain model that's in BISON. An obvious reason for it diffing is not jumping out at me.

**Comment by WilkAndy** [comment]
That's good to know, thanks @bwspenc .  Now i know where to concentrate my efforts (ComputeFiniteStrainElasticStress).  I'll use Bison as a black box and trial things during today:  there aren't many changes in this PR so it should be possible to find the error.

**Comment by bwspenc** [comment]
I was going to mention that I just noticed today that some of those same tests are giving us trouble on our compute cluster. I wonder if they're just a little flaky. It looks like you have them passing now, though, so maybe you found a solution.

**Comment by sapitts** [comment]
@WilkAndy sorry for my delay on looking at these Bison tests and on reviewing the code.  I'll pull your branch done and see if I can help make Bison less black box here shortly. From @bwspenc it sounds like I might need to patch these tests.

We need the isotropic check in Bison (and in Moose!) to ensure that we calculate the old stress from the old strain correctly.  With an anisotropic elasticity tensor the inversion doesn't work correctly, I think because of rotations, but I'll have to dig out my notes from when @bwspenc and I talked about this.  I'll focus on getting those tests fixed first though!

**Comment by sapitts** [comment]
@WilkAndy as an in progress report, two of the failing tests are thermal expansion tests.  These are failing with dtmin when the temperature is held constant at the end of a simulation for a couple of extra timesteps; the timestep when delta Temperature is zero is when the tests fail in dtmin.  I'll work on this more (my) tomorrow

**Comment by WilkAndy** [comment]
Cool, thanks @sapitts

**Comment by WilkAndy** [comment]
I reviewed my changes and changed ComputeFiniteStrainElasticStress back to its previous version, but still BISON crapped out.  I don't understand what the problem is.

**Comment by sapitts** [comment]
Of the four tests failing, two (the thermal expansion) used the computefinitestrainelasticstress and two (creep tests) used the ComputeMultipleInelasticStress; the latter two are still failing.  I'd like to get all four tests passing and without the need for an incompatible API update

**Comment by WilkAndy** [comment]
Thanks @sapitts .  I'd like to get to the bottom of why those tests are failing.  I don't see why my changes should effect anything at all - according to me they are just a small refactor of the initial_stress stuff, but should give identical results to the previous version.  So obviously i'm not understanding something, and if you have the time I'd greatly appreciate being educated as to what my oversight is.  (This current version is desirable for me because it allows initial stresses to be prescribed even when the elasticity tensor is changing.)

a

**Comment by sapitts** [comment]
I agree @WilkAndy; the creep tests can probably be solved by raising the floor, and then I can regold once your commit is through.  The thermal expansion tests are more puzzling though,and I'll take another look this afternoon, fingers crossed.

**Comment by WilkAndy** [comment]
Update on this initial_stress stuff.  I found out that for my StressUpdate classes that
```
_stress_old[_qp] - _elasticity_tensor[_qp] * _elastic_strain_old[_qp]
```
was not zero!  This was because of tolerances in my return-mapping algorithm not being set tight enough, and resetting them made the above equal to zero.  In my return-map algorithm, i compute the increment of inelastic strain from looking at the returned stress and the trial stress, but NOT using the inverse of the elasticity tensor.  I bet you do the same, because of the computational expense.  Errors were accruing, and because the above wasn't zero, the following two gave different results:
```
   _stress[_qp] = _elasticity_tensor[_qp] * (_elastic_strain_old[_qp] + elastic_strain_increment);
    addQpInitialStress(_stress[_qp]);
```
and
```
    _stress[_qp] = _stress_old[_qp] + _elasticity_tensor[_qp] * elastic_strain_increment;
```

I wonder if the same problem occurs for you?

I still think that if the above two gave the same result for stress then the changes in this PR shouldn't effect any outputs.

**Comment by WilkAndy** [comment]
Pushed the "close" button by accident!

**Comment by WilkAndy** [comment]
Any further progress on this, @sapitts ?

**Comment by sapitts** [comment]
Nothing I've pushed yet, but it's at the top of my list for tomorrow.  Sorry for holding up your commit with the Bison tests.

**Comment by WilkAndy** [comment]
@sapitts - i'm just so pleased to work collaboratively with you on this!  I don't have to figure out everything myself :-)

**Comment by WilkAndy** [comment]
Oh dear, i think this is causing you headaches, sorry @sapitts  :-(

**Comment by sapitts** [comment]
@WilkAndy I've got a patch for Bison that I'll put in for review shortly.  I couldn't correctly fix the creep tests that were failing, so I've used a not ideal technique of loosing tolerances until your merge request goes though (when I'll regold and tighten tolerances again)

**Comment by WilkAndy** [comment]
Oh, that's disappointing and i feel bad that i mucked you around so much, sorry.  But thanks so much for working on this.

a

**Comment by sapitts** [comment]
@WilkAndy we got the patch into Bison and it's making it's way through to the Bison master branch so that we can get your pull request passing the testing this afternoon :-)

**Comment by WilkAndy** [comment]
Thanks @sapitts .   It's such a pleasure working with you INL people.

a

File: modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C
Comment (sapitts) [inline modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C]: Do we still need to check for if the elasticity tensor is isotropic?

File: modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C
Comment (sapitts) [inline modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C]: Is there a potential problem with keeping the initial stress in the same current configuration with the stress?  The initial stress is always given in the reference configuration, but we only apply the incremental rotation to this initial stress quantity.

**Review by sapitts — COMMENTED** [review]
Straight forward change. I'm looking forward to your thoughts about incremental rotations and the initial stress

**Comment by WilkAndy** [comment]
Good points, thanks @sapitts.  I'll address them in a few hours.

File: modules/tensor_mechanics/src/materials/ComputeStressBase.C
Comment (WilkAndy) [reply modules/tensor_mechanics/src/materials/ComputeStressBase.C]: I see that RankTwoTensor has a private static constexpr unsigned N=LIBMESH_DIM.  So we should use LIBMESH_DIM instead of 3, for consistency.  Perhaps we could make that "N" parameter public and use that instead, for ease of changing later on.  But that's for another PR.

File: modules/tensor_mechanics/include/materials/ComputeStressBase.h
Comment (WilkAndy) [reply modules/tensor_mechanics/include/materials/ComputeStressBase.h]: clarified doco, thanks

File: modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C
Comment (WilkAndy) [reply modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C]: OK, i have completely re-written this stuff because of the finite-strain rotations.  I've tried to have no performance or memory penalties in the case where initial_stress is not given, because i think that's the most common use-case in MOOSE

File: modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C
Comment (WilkAndy) [reply modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C]: Probably we need the check for isotropic elasticity tensor.  It wasn't there originally, but i'll include it in the current version.

Comment (WilkAndy) [reply modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C]: Oh, i take that back.  ComputeFiniteStrainElasticStress can only be used for isotropic tensors - there's a check in initialSetup that throws a mooseError otherwise.

**Comment by WilkAndy** [comment]
This PR will probably pass when the clang-formatting is fixed, but **please don't merge yet**.   I want to experiment with initial eigenstrains instead of the initial stress.  I think it might make for cleaner code.

**Comment by WilkAndy** [comment]
I'm happy with this.  It is similar to what you've seen before, @sapitts , with a lot of the constant_in_time stuff removed, but now includes an Eigenstrain calculator (ComputeEigenstrainFromInitialStress) because i think that's the way i should be including initial stresses, as you stated quite a few months ago, sigh!

I'm going to remove initial_stress from ComputeStressBase on 1March2018.  It's now a deprecated parameter.

Refs #10074

**Comment by WilkAndy** [comment]
Bummer.... days waiting for my PR to be tested and then it fails because i forgot to include a couple of new test input files.

**Comment by WilkAndy** [comment]
Modules failed, probably due to #10085

Mastodon failed, due to the "deprecated" message i included for the initial_stress parameter.  I can't remember who maintains Mastodon.  This PR is taking so far longer to get through than I had imagined, so i'm going to remove the "deprecated" message and include it in a different PR later.

a

**Comment by WilkAndy** [comment]
@sapitts - is this "initial_stress" PR ready to merge now?

File: modules/tensor_mechanics/src/materials/ComputeMultipleInelasticStress.C
Comment (bwspenc) [inline modules/tensor_mechanics/src/materials/ComputeMultipleInelasticStress.C]: It doesn't like you're using `_is_elasticity_tensor_guaranteed_constant_in_time` anymore in this class.  Can you get rid of it?

Comment (bwspenc) [reply modules/tensor_mechanics/src/materials/ComputeMultipleInelasticStress.C]: Along those lines, does this PR eliminate all usages of CONSTANT_IN_TIME? If so, you should remove it as well.

Comment (WilkAndy) [reply modules/tensor_mechanics/src/materials/ComputeMultipleInelasticStress.C]: Re eliminating all constant_in_time stuff - yes, maybe.  I'll change the PR and see what happens.  I suspect that dependent apps might use the const_in_time - we'll see.

File: modules/tensor_mechanics/src/materials/ComputeEigenstrainFromInitialStress.C
Comment (sapitts) [inline modules/tensor_mechanics/src/materials/ComputeEigenstrainFromInitialStress.C]: I'm mildly concerned about what happens if the elasticity tensor changes a great deal in the first timestep, say with a large temperature spike or lots of damage, but I'm not sure how relevant of a corner case this is.  Perhaps it might be worth running the calculation in initialQpProperties but then saving the eigenstrain value calculated in `_t_step --1`?  Or maybe a message in the class description to use care if the elasticity tensor changes drastically at the simulation start?

File: modules/tensor_mechanics/src/materials/ComputeStressBase.C
Comment (sapitts) [inline modules/tensor_mechanics/src/materials/ComputeStressBase.C]: Mark the initial stress parameter above as deprecated?

**Review by sapitts — COMMENTED** [review]
Looks pretty good to me @WilkAndy! The additional example input would be nice in the documentation page, and everything else is pretty minor

File: modules/tensor_mechanics/src/materials/ComputeEigenstrainFromInitialStress.C
Comment (WilkAndy) [reply modules/tensor_mechanics/src/materials/ComputeEigenstrainFromInitialStress.C]: I toyed with the idea of running the eigenstrain calculation in initQp (for other reasons actually).  It won't work without a lot of bother, since the elasticity tensor is not defined then (it's not stateful).  I don't think it's worth going to a great deal of effort at the moment before we really have a use-case where this matters.  I'll certainly include a message in the doco string - thanks.

If you can see a better way of doing this, please tell me!

**Comment by WilkAndy** [comment]
I'll address @sapitts comments after checking whether completely removing constant_in_time causes any breakages.

File: docs/content/documentation/systems/Materials/tensor_mechanics/ComputeEigenstrainFromInitialStress.md
Comment (WilkAndy) [reply docs/content/documentation/systems/Materials/tensor_mechanics/ComputeEigenstrainFromInitialStress.md]: done

File: docs/content/documentation/systems/Materials/tensor_mechanics/ComputeEigenstrainFromInitialStress.md
Comment (WilkAndy) [reply docs/content/documentation/systems/Materials/tensor_mechanics/ComputeEigenstrainFromInitialStress.md]: done

File: modules/tensor_mechanics/src/materials/ComputeCosseratLinearElasticStress.C
Comment (WilkAndy) [reply modules/tensor_mechanics/src/materials/ComputeCosseratLinearElasticStress.C]: Done (changed TODO to Deprecation)

File: modules/tensor_mechanics/src/materials/ComputeEigenstrainFromInitialStress.C
Comment (WilkAndy) [reply modules/tensor_mechanics/src/materials/ComputeEigenstrainFromInitialStress.C]: I added a big note regarding this to the .md file

File: modules/tensor_mechanics/src/materials/ComputeStressBase.C
Comment (WilkAndy) [reply modules/tensor_mechanics/src/materials/ComputeStressBase.C]: Yep, i'm going to mark the initial stress parameter as deprecated, but can't do that until this PR gets merged, then i can change Mastodon, and then change to the deprecatedParam

File: modules/tensor_mechanics/src/materials/ComputeStressBase.C
Comment (WilkAndy) [reply modules/tensor_mechanics/src/materials/ComputeStressBase.C]: Yea, i wonder how many other "3"s there are hanging around!

File: modules/tensor_mechanics/test/tests/initial_stress/except01.i
Comment (WilkAndy) [reply modules/tensor_mechanics/test/tests/initial_stress/except01.i]: done

**Comment by WilkAndy** [comment]
Finally!   I hope you can merge this soon, hint hint!

**Review by sapitts — APPROVED** [review]
Looks good to me!  Thanks @WilkAndy

File: modules/tensor_mechanics/src/materials/ComputeEigenstrainFromInitialStress.C
Comment (bwspenc) [inline modules/tensor_mechanics/src/materials/ComputeEigenstrainFromInitialStress.C]: I've been wanting to get rid of `_incremental_form` for some time.  There are only a couple of models that use it. I think this is the right thing to do for now, but when we remove `_incremental_form`, we'll just have to have this model get the old eigenstrain on its own.

File: modules/tensor_mechanics/src/materials/ComputeStressBase.C
Comment (bwspenc) [reply modules/tensor_mechanics/src/materials/ComputeStressBase.C]: This is a discussion for another time, but I think `LIBMESH_DIM` is horrible to use for 3, because it can be set to 1, 2, or 3 at compile time.  We need some other tensor size thing that is always 3. We'll just do a big search/replace on `LIBMESH_DIM` if we ever get around to it, so this is fine for now.

File: modules/tensor_mechanics/src/materials/ComputeEigenstrainFromInitialStress.C
Comment (bwspenc) [reply modules/tensor_mechanics/src/materials/ComputeEigenstrainFromInitialStress.C]: I guess the concern here is if the tensor changes between time 0 and time 1.  Let's go ahead and merge this code as it is, but I'd like to see if there's some way we can figure out how to compute the eigenstrain using the elasticity tensor as it is at time 0, so that it would always do the right thing.

**Review by bwspenc — APPROVED** [review]

**Comment by WilkAndy** [comment]
Thanks for all your comments and most of all your patience with this PR, @sapitts and @bwspenc .   It's an important feature for me to have.

a

**Pull request merged**
